### DISCUSSION TITLE
How would you implement a Mixin with pyo3

### DISCUSSION THREAD

**Discussion by llacroix** [PyO3/pyo3#3563] [Questions] [answered]
Let say I have a class named:

```
class A(object):
    def hello(self):
        return " world!"
```

And a python module using the rust mixin:

```
import rust_module

class B(rust_module.Mixin, A):
    pass

b = B()
b.hello()
>>> "hello world!"
```
 
Basically it would need to call super on the parent class, but rust doesn't already know which parent is super at compile time. 

I tried using 

```
#[new]
fn new(self_: &PyCell<Self>) -> PyResult(&PyAny) {
   Ok(Mixin {})
}
```

But it's complaining about Mixin.__new__() missing 1 required positional argument self_

 With:

```
#[new]
fn new() -> Self {
  Mixin {}
}
```

It doesn't seem like the parent class is properly set as new fails with recursion error when calling yell using PyCell  . py_super

``` 
use pyo3::prelude::*;

#[pyclass(subclass)]
struct Mixin {}

#[pymethods]
impl Mixin {
  #[new]
  fn new() -> Self {
    Mixin {}
  }

  pub fn hello(self_: &PyCell<Self>) -> PyResult<&PyAny> {
    let super_ = self_.py_super()?;
    super_.call_method("hello", (), None)
  }
}

#[pymodule]
fn rust_module(_py: Python<'_>, m: &PyModule) -> PyResult<()> {
  m.add_class::<Mixin>()?;
  Ok(())
}
```

**Comment by davidhewitt** [comment, marked as answer]
Try 

```rust
let super_ = PySuper:: new(Mixin::type_object(py), self_)?;
```

... I suspect that `py_super` is probably only looking at the type `B`, which is why you get the recursion error. Probably a design oversight.

**Reply by llacroix** [reply]
Perfect, it works with:

```
let super_ = PySuper::new(Mixin::type_object(self_.py()), self_)?;
super_.call_method('hello', (), None)
```
